Move C++17.h to ATen/core

smessmer opened the pull request:
Differential Revision: D9095209

ezyang reviewed (approved):
OK, but it's gonna conflict with #10092

smessmer commented:
How is this going to conflict? I thought you said you didn't move C++17.h?

smessmer committed:
Move C++17.h to ATen/core

Summary:
This header is needed for ATen/core stuff

This diff also fixes an issue in C++17.h when run in C++17 enabled compilers.

Differential Revision: D9095209

fbshipit-source-id: b059f0a1f8be160ce48c51c3c2c195af525de474

smessmer commented:
ok somehow shipit didn't connect this to the diff and now crashes when I try to manually do it. Closing and reopening.
